src/libraries/System.Net.Security/src/System/Net/Security/SslStream.IO.cs

```diff
@@ -120,9 +125,29 @@ private Task ProcessAuthenticationAsync(bool isAsync = false, CancellationToken
 
         private async Task ProcessAuthenticationWithTelemetryAsync(bool isAsync, CancellationToken cancellationToken)
         {
-            NetSecurityTelemetry.Log.HandshakeStart(IsServer, _sslAuthenticationOptions.TargetHost);
-            long startingTimestamp = Stopwatch.GetTimestamp();
+            long startingTimestamp;
+            if (NetSecurityTelemetry.Log.IsEnabled())
+            {
+                NetSecurityTelemetry.Log.HandshakeStart(IsServer, _sslAuthenticationOptions.TargetHost);
+                startingTimestamp = Stopwatch.GetTimestamp();
+            }
+            else
+            {
+                startingTimestamp = 0;
+            }
 
+            using Activity? activity = s_activitySource.StartActivity(ActivityName, IsServer ? ActivityKind.Server : ActivityKind.Client);
+            if (activity is not null)
+            {
+                activity.DisplayName = IsServer ? "TLS server" : $"TLS client {TargetHostName}";
+                if (activity.IsAllDataRequested && !IsServer)
+                {
+                    activity.SetTag("tls.client.server_name", TargetHostName);
+                }
```

Comment:
This will be renamed to `server.address`, see https://github.com/open-telemetry/semantic-conventions/pull/1192#discussion_r1667406120.

> I assume we could add a tag when we receive one from the client

This is being set for the client only.